fqdn proxy: fix data race by using separate sessionUDPFactories

mhofstetter opened the pull request:
PR #25309 introduced a data race by sharing the sessionUDPFactory between the DNS server instances for the different IP families (IPv4 & IPv6). This has been detected by #27979.

This commit fixes the data race by using dedicated instances of the sessionUDPFactory.

In addition the global response UDP connections (IPv4 & IPv6) variables have been replaced with fields in the `SessionUDPFactory`. This way the connections can already be setup during construction of the factory and the `sync.Once`s (global vars) can be removed.

Fixes: #28156

mhofstetter committed:
fqdn proxy: fix data race by using separate sessionUDPFactories

PR #25309 introduced a data race by sharing the sessionUDPFactory between the
DNS server instances for the different IP families (IPv4 & IPv6). This has been detected
by #27979.

This commit fixes the data race by using dedicated instances of the sessionUDPFactory.

Fixes: #28156

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
fqdn proxy: replace global response conn with field in SessionUDPFactory

This commit replaces the global response UDP connections (IPv4 & IPv6)
variables with fields in the SessionUDPFactory. This way the connections
can already be setup during construction of the factory and the `sync.Once`s
(global vars) can be removed.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -30,14 +31,48 @@ var udpOOBSize = func() int {
-	ipv4Enabled, ipv6Enabled bool
+	// responseConn4 and responseConn6 are used to send the response
+	// See sessionUDP.WriteResponse
+	responseConn4, responseConn6 *net.IPConn
```
How about keeping only one pointer (`responseConn`) in the factory? Each factory instance only has one family defined, so we do not need two pointers here?

mhofstetter replied:
The reason is that the "building blocks" (e.g. SessionUDPFactory, setting transparent socket options, ...) were all built when a socket bound to a Cilium DNS server supported multiple IP Famliies (IPv4 & IPv6).

But this is no longer necessary since we use a dedicated DNS Server (& socket) per IPFamily.

Therefore i agree that this could be simplified. But for all cases.

=> I refactored the SessionUDPFactory & "setting the transparent socket options" to only support one IPFamily. All the relevant information are read from the `IPFamily` struct that gets passed down - to keep the code agnostic of the actual IP Family.

jrajahalme reviewed (commented):
Could simplify a bit by keeping only one conn pointer in the factory.

mhofstetter committed:
fqdn proxy: binding socketUDPFactory to specific IP family

Historically, the `socketUDPFactory` and the lilsten configuration
supported IPv4 and IPv6 on the same socket.

Since the change to bind the DNS servers to localhost, a server
and socket is created per ip family.

Therefore this commit removes the support for multiple ipfamilies
on the same socket (SocketUDPFactory & listen configuration). This
improves readability quite a lot.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
fqdn proxy: add transparent socket options to IPFamily struct

This commit adds the transparent socket options to the IPFamily
struct. This way it can be used in the respective functions.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme commented on pkg/fqdn/proxy/ipfamily/ipfamily.go:
```diff
@@ -3,33 +3,41 @@
+
+	SocketOptsFamily              int
+	SocketOptsTransparent         int
+	SocketOptsRecvOrIgnoreDstAddr int
```
This should be `SocketOptsRecvOrigDstAddr` (receive original destination address)

mhofstetter replied:
the "power" of autocompletion :joy_cat: addressed and fixed - thanks!

jrajahalme reviewed (approved):
One nit I hope you address.

pippolo84 commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -51,41 +70,29 @@ type sessionUDP struct {
+// By design, a socket of a DNS Server can only receive IPv4 or IPv6 traffic.
+func transparentSetsockopt(fd int, ipFamily ipfamily.IPFamily) error {
+	if err := unix.SetsockoptInt(fd, ipFamily.SocketOptsFamily, ipFamily.SocketOptsTransparent, 1); err != nil {
+		return fmt.Errorf("setsockopt(IP(V6)_TRANSPARENT) failed: %w", err)
```
i think the error string should contain `IPv6` or `IPv4` based on `ipFamily`

mhofstetter replied:
thanks, i changed the error message to contain the name of the `ipFamily`

pippolo84 commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -51,41 +70,29 @@ type sessionUDP struct {
-			return fmt.Errorf("setsockopt(IP_RECVORIGDSTADDR) failed: %w", err)
-		}
+	if err := unix.SetsockoptInt(fd, ipFamily.SocketOptsFamily, ipFamily.SocketOptsRecvOrigDstAddr, 1); err != nil {
+		return fmt.Errorf("setsockopt(IP(V6)_RECVORIGDSTADDR) failed: %w", err)
```
Same here

pippolo84 reviewed (commented):
Just a minor thing related to an error string, apart from that LGTM!

pippolo84 reviewed (approved):
:100:
